demos/instr.h

```diff
@@ -157,6 +158,30 @@ int ExchangeFS(int input);
 // Returns the original value of ES and sets the new value.
 int ExchangeES(int input);
 
+// Performs a bound check with the bound instruction. Works only on 32-bit x86.
+// Must be inlined in order to avoid mispredicted jumps over it.
+SAFESIDE_ALWAYS_INLINE
+inline void BoundsCheck(const char *str, size_t offset) {
+  struct {
+    int32_t low;
+    int32_t high;
+  } string_bounds;
+
+  string_bounds.low = 0;
+  string_bounds.high = strlen(str);
```

Comment:
The documentation on the `BOUND` instruction has this weird phrasing:
> The array index must be greater than or equal to the lower bound and less than or equal to the upper bound plus the operand size in bytes

I'm not sure exactly how to read that, but the pseudocode is:
```
IF (ArrayIndex < LowerBound OR ArrayIndex > UpperBound) THEN
(* Below lower bound or above upper bound *)
  IF <equation for PL enabled> THEN
    BNDSTATUS ← 0
    #BR;
  FI;
```

and so I think the upper bound here should be `strlen(str) - 1`.

Reply:
That's correct. Fixed.